File: insert-interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<Interval> result = new LinkedList<Interval>();
        int i = 0;
        while (i < intervals.length && intervals[i][1] < newInterval[0]) {
            result.add(new Interval(intervals[i]));
            i++;
        }
        
        int start = newInterval[0];
        int end = newInterval[1];
        
        while (i < intervals.length && intervals[i][0] <= newInterval[1]) {
            start = Math.min(start, intervals[i][0]);
            end = Math.max(end, intervals[i][1]);
            i++;
        }
        
        result.add(new Interval(start, end));
        
        while (i < intervals.length) {
            result.add(new Interval(intervals[i]));
            i++;
        }
        
        int[][] ans = new int[result.size()][2];
        int j = 0;
        for (Interval interval : result) {
            ans[j][0] = interval.start;
            ans[j][1] = interval.end;
            j++;
        }
        return ans;
    }
    
}
​
class Interval {
    int start;
    int end;
    
    Interval(int start, int end) {
        this.start = start;
        this.end = end;
    }
    
    Interval(int[] interval) {
        this.start = interval[0];
        this.end = interval[1];
    }
}

File: insert-interval/insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        ans = []
        
        i = 0
        while i < len(intervals):
            if newInterval[0] < intervals[i][0]:
                break
            
            ans.append(intervals[i])
            i += 1
        
        # Insert newInterval
        if len(ans) == 0 or not self.is_overlap(ans[-1], newInterval):
            ans.append(newInterval)
        else:
            ans[-1] = self.merge(ans[-1], newInterval)
        
        while i < len(intervals):
            if self.is_overlap(ans[-1], intervals[i]):
                ans[-1] = self.merge(ans[-1], intervals[i])
            else:
                ans.append(intervals[i])
            i += 1
            
        return ans
    
    def is_overlap(self, i1, i2):
        return i1[0] <= i2[1] and i1[1] >= i2[0]
    
    def merge(self, i1, i2):
        if not self.is_overlap(i1, i2):
            return None
        
        start = min(i1[0], i2[0])
        end = max(i1[1], i2[1])
        return [start, end]
        